core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java

```diff
@@ -67,18 +71,36 @@ public SuggestContextParser getContextParser() {
             // if we index a suggestion with n contexts, the suggestion and all its contexts
             // would count as n hits rather than 1, so we have to multiply the desired size
             // with n to get a suggestion with all n contexts
-            final String key = suggestDoc.key.toString();
             final float score = suggestDoc.score;
             final Map.Entry<String, CharSequence> contextEntry;
-            if (suggestionContext.fieldType().hasContextMappings() && suggestDoc.context != null) {
-                contextEntry = suggestionContext.fieldType().getContextMappings().getNamedContext(suggestDoc.context);
+            if (suggestionContext.getFieldType().hasContextMappings() && suggestDoc.context != null) {
+                contextEntry = suggestionContext.getFieldType().getContextMappings().getNamedContext(suggestDoc.context);
             } else {
                 assert suggestDoc.context == null;
                 contextEntry = null;
             }
-            final Option value = results.get(suggestDoc.doc);
+            final CompletionSuggestion.Entry.Option value = results.get(suggestDoc.doc);
             if (value == null) {
-                final Option option = new Option(suggestDoc.doc, new StringText(key), score, contextEntry);
+                final Map<String, List<Object>> payload;
+                Set<String> payloadFields = suggestionContext.getPayloadFields();
+                if (!payloadFields.isEmpty()) {
+                    int readerIndex = ReaderUtil.subIndex(suggestDoc.doc, searcher.getIndexReader().leaves());
+                    LeafReaderContext subReaderContext = searcher.getIndexReader().leaves().get(readerIndex);
+                    int subDocId = suggestDoc.doc - subReaderContext.docBase;
+                    payload = new LinkedHashMap<>(payloadFields.size());
+                    for (String field : payloadFields) {
+                        MappedFieldType fieldType = suggestionContext.getMapperService().smartNameFieldType(field);
+                        if (fieldType != null) {
```

Comment:
Hmm, if `fieldType` is null it means the use asked for a non-existent payloads field?  Can we throw an exc (earlier) so user realizes this?